### ISSUE TITLE
caffe 精度损失，尝试AGW_R18转onnx 与caffe，其中onnx 特征结果接近 torch，但是caffe差异比较大

### ISSUE THREAD

**Issue by hp-93** [JDAI-CV/fast-reid#274]
AGW_R18x  训练config：
_BASE_: "../Base-AGW.yml"
MODEL:
  BACKBONE:
    DEPTH: "18x"
    WITH_IBN: False
    WITH_NL: False
  HEADS:
    IN_FEAT: 512

DATASETS:
  NAMES: ("xxxx2020",)
  TESTS: ("xxxx2020",)

INPUT:
  SIZE_TRAIN: [128, 128]
  SIZE_TEST: [128, 128]

TEST:
  EVAL_PERIOD: 10
  IMS_PER_BATCH: 1

OUTPUT_DIR: "logs/xxxx/agw_R18"

caffe 差异：
```
<     ceil_mode: false
---
>     #ceil_mode: false
941a942,953
>
> layer {
>     name: "avgpool1"
>     type: "Pooling"
>     bottom: "relu_blob17"
>     top: "avgpool_blob1"
>     pooling_param {
>         pool: AVE
>         global_pooling: true
>     }
> }
>
945c957
<   bottom: "relu_blob17"
---
>   bottom: "avgpool_blob1"
956c968
<   top: "batch_norm_blob21"
---
>   top: "output"
```

onnx 单个图像结果：
outputs tensor([[-3.0259e-01,  1.0013e-01, -2.4013e-01,  1.8437e-02,  2.2592e-01,
         -1.3462e-01, -2.6042e-01,  2.2218e-01,  9.8405e-01,  8.5403e-03,....]])

cafffe 结果 ：
[[ -5.44965148e-01  -3.35639596e-01  -9.01259065e-01  -2.61804998e-01
   -3.98053318e-01  -5.82201898e-01  -3.62800360e-01  -1.00999460e-01
   -1.66855946e-01  -6.58459306e-01  -7.58686304e-01  -6.80805027e-01
   -5.46413064e-01  -5.07166743e-01  -2.92615652e-01  -9.08304632e-01
   -6.19151771e-01   3.48099135e-02  -9.05649006e-01  -8.19128931e-01]]

**Comment by hp-93** [comment]
caffe 预处理去掉了减均值 
![image](https://user-images.githubusercontent.com/35527573/93575137-d676ab00-f9cb-11ea-8561-95cc588ef60a.png)

**Comment by hp-93** [comment]
加上image = (image - np.array([0.485 * 255, 0.456 * 255, 0.406 * 255]).reshape((1, -1, 1, 1))) / np.array(
        [0.229 * 255, 0.224 * 255, 0.225 * 255]).reshape((1, -1, 1, 1))
![image](https://user-images.githubusercontent.com/35527573/93581915-599bff00-f9d4-11ea-888f-ed8cd640910d.png)

**Comment by hp-93** [comment]
平均池化层输入维度不太对应，onnx 对应是(1, 512, 8, 8)， caffe 对应 (1, 512, 9, 9)

**Comment by L1aoXingyu** [comment]
https://github.com/JDAI-CV/fast-reid/issues/210#issuecomment-669010431

可以参考这个 issue

**Comment by hp-93** [comment]
我试过maxpool 换成self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, ceil_mode=True)， 对比GlobalAveragePool 层之后结果：![image](https://user-images.githubusercontent.com/35527573/93657011-484d0400-fa61-11ea-9239-7ec38cf4ff93.png)
维度是对应的，结果大概也是对的
caffe
![image](https://user-images.githubusercontent.com/35527573/93657035-86e2be80-fa61-11ea-9703-e1c7ce47ecbc.png)
onnx
![image](https://user-images.githubusercontent.com/35527573/93657026-6f0b3a80-fa61-11ea-9d27-c491be146db7.png)
但是 加上POW 与 BatchNormalization就不太对了
![image](https://user-images.githubusercontent.com/35527573/93657045-a843aa80-fa61-11ea-911b-c24d4e2e97fd.png)
![image](https://user-images.githubusercontent.com/35527573/93657048-af6ab880-fa61-11ea-8674-cce0f9d68236.png)

**Comment by hp-93** [comment]
比较torch， onnx 与 caffe 最后一层特征结果为：
torch：
![image](https://user-images.githubusercontent.com/35527573/93657168-cb228e80-fa62-11ea-89f4-ec7963ad31ea.png)
onnx
![image](https://user-images.githubusercontent.com/35527573/93657180-daa1d780-fa62-11ea-9403-443a5ebf2929.png)
caffe
![image](https://user-images.githubusercontent.com/35527573/93657183-e1c8e580-fa62-11ea-8625-71b44db5d714.png)

**Comment by hp-93** [comment]
放弃使用caffe，直接用onnx

**Comment by sky186** [comment]
@hp-93  
您好，我像问您一下 mgn _ibn 50  pytorch  to onnx    ，   clip  和 pow layer   是pytorch 哪个层转换出来？，  转换到caffe 这两层不知道怎么处理？      为什么放弃caffe ,直接用onnx ,ibn 层应该过不去，    所以您现在转成tensorRT 了吗？
